src/box/lua/console.c

```diff
@@ -266,6 +267,124 @@ console_sigint_handler(ev_loop *loop, struct ev_signal *w, int revents)
 	fiber_wakeup(interactive_fb);
 }
 
+/* {{{ Show/hide prompt */
+
+/*
+ * The idea is borrowed from
+ * https://metacpan.org/dist/AnyEvent-ReadLine-Gnu/source/Gnu.pm
+ */
+
+static char *saved_prompt = NULL;
+static char *saved_line_buffer = NULL;
+static int saved_line_buffer_len = 0;
+static int saved_point = 0;
+
+static int console_hide_prompt_ref = LUA_NOREF;
+static int console_show_prompt_ref = LUA_NOREF;
+
+static void
+console_hide_prompt(void)
+{
+	if (rl_prompt == NULL) {
```

Comment:
If we'll not find a good solution for the case, when `rl_message()` is called (usually at `Ctrl+R`, but also at `M-<number>`), it is better to don't try to save/restore the prompt at all. Things becomes even worse if we'll doing so, because readline performs its own save/restore actions and expects that the prompt is not changed.

We can check `RL_ISSTATE(RL_STATE_ISEARCH|RL_STATE_NUMERICARG)`.

Reply:
I disabled the prompt hide/show code in states, where `rl_message()` is called by readline.

I made several attempts to handle this case, but didn't find a solution.